source/array_list.c

```diff
@@ -0,0 +1,247 @@
+    assert(array_size >= item_size);
+
+    list->alloc = NULL;
+
+    size_t size_ratio = array_size % item_size;
+    if(size_ratio != 0) {
+        assert(0);
+        return aws_raise_error(AWS_ERROR_LIST_STATIC_ILLEGAL_SIZE);
+    }
+
+    list->current_size = array_size;
+    list->item_size = item_size;
+    list->length = 0;
+    list->data = raw_array;
+    return AWS_OP_SUCCESS;
+}
+
+void aws_array_list_clean_up(struct aws_array_list *list) {
+    if (list->alloc && list->data) {
+        aws_mem_release(list->alloc, list->data);
+    }
+
+    list->current_size = 0;
+    list->item_size = 0;
+    list->length = 0;
+    list->data = NULL;
+    list->alloc = 0;
+}
+
+int aws_array_list_push_back(struct aws_array_list *list, const void *val) {
+
+    size_t filled_space = list->length  * list->item_size;
+    if (filled_space == list->current_size) {
+        if (list->alloc) {
+            size_t new_size = list->current_size << 1;
+            if (new_size == 0) {
+                new_size = 2;
+            }
+
+            if (new_size < list->current_size) {
+                /* this means new_size overflowed. The only way this happens is on a 32-bit system
+                 * where size_t is 32 bits, in which case we're out of addressable memory anyways, or
+                 * we're on a 64 bit system and we're most certainly out of addressable memory.
+                 * But since we're simply going to fail fast and say, sorry can't do it, we'll just tell
+                 * the user they can't grow the list anymore. */
+                return aws_raise_error(AWS_ERROR_LIST_EXCEEDS_MAX_SIZE);
+            }
+
+            void *temp = aws_mem_acquire(list->alloc, new_size);
+
+            if (!temp) {
+                return aws_raise_error(AWS_ERROR_OOM);
+            }
+
+            memset((uint8_t *)temp + list->length, 0, new_size - list->current_size);
+
+            if (list->data) {
+                memcpy(temp, list->data, list->current_size);
+                aws_mem_release(list->alloc, list->data);
+            }
+
+            list->data = temp;
+            list->current_size = new_size;
+        }
+        else {
+            return aws_raise_error(AWS_ERROR_LIST_EXCEEDS_MAX_SIZE);
+        }
+    }
+
+    memcpy((void *)((uint8_t *)list->data + filled_space), val, list->item_size);
+    list->length++;
+    return AWS_OP_SUCCESS;
+}
+
+int aws_array_list_front(const struct aws_array_list *list, void *val) {
+
+    if (list->length > 0) {
+        memcpy(val, list->data, list->item_size);
+        return AWS_OP_SUCCESS;
+    }
+
+    return aws_raise_error(AWS_ERROR_LIST_EMPTY);
+}
+
+int aws_array_list_pop_front(struct aws_array_list *list) {
+    if (list->length > 0) {
+        size_t last_bytes = list->item_size * (list->length - 1);
+        memmove(list->data, (void *)((uint8_t *)list->data + list->item_size), last_bytes);
+        memset((void *)((uint8_t *)list->data + last_bytes), 0, list->item_size);
```

Comment:
I like debuggable code though. Having a set of 0s in unused space makes tracking down the bugs easier, but it's also a performance penalty, maybe I'll ifdef this off with the debug flag.

Reply:
It can be more helpful to use a nonzero sentinel value (e.g. `0xDD`) or even a multi-byte pattern (`0xDEADBEEF`) to mark uninitialized memory - the reason being, often _initialized_ structures are all-zero.

Reply:
Good thought, I'll change, and still have it sitting behind a DEBUG flag.